ui/src/app/store/machine/utils/common.ts

```diff
@@ -9,3 +10,24 @@ export const isMachineDetails = (
   machine?: Machine | null
   // Use "metadata" as the canary as it only exists for MachineDetails.
 ): machine is MachineDetails => !!machine && "metadata" in machine;
+
+export type TagIdCountMap = Map<Tag[TagMeta.PK], number>;
+
+/**
+ * The tag ids for the given machines.
+ * @param machines - The machines to get the tag ids from.
+ * @returns A list of tag ids.
+ */
+export const getTagCountsForMachines = (machines: Machine[]): TagIdCountMap => {
+  const ids = machines.reduce<Tag[TagMeta.PK][]>(
+    (tagIds, machine) => tagIds.concat(machine.tags),
+    []
+  );
+  const tagCounts = new Map();
+  ids.forEach((id) => {
+    if (!tagCounts.has(id)) {
+      tagCounts.set(id, ids.filter((tagId) => tagId === id).length);
+    }
```

Comment:
It seems a bit inefficient to filter the entire tag ID list on each new entry. Would a simple conditional work here?

``` ts
tagCounts.set(id, (tagCounts.get(id) || 0) + 1);
```

Reply:
Then again I dunno if continuously updating a map entry is a more expensive operation, so maybe the point is moot

Reply:
I just ran both versions using the profiler and with my dataset on my machine the `+ 1` version was on average 0.1ms slower, but we're talking an imperceptibly small amount of difference. It's possible a different data set could return different results, but I guess we can revisit this if we're noticing performance issues with large amounts of data.